presto-orc/src/test/java/com/facebook/presto/orc/TestCachingOrcDataSource.java

```diff
@@ -0,0 +1,298 @@
+        cachingOrcDataSource.readCacheAt(3);
+        assertEquals(delegateOrcDataSource.getLastReadRanges(), ImmutableList.of(new DiskRange(3, 60)));
+        cachingOrcDataSource.readCacheAt(63);
+        assertEquals(delegateOrcDataSource.getLastReadRanges(), ImmutableList.of(new DiskRange(63, 8 * 1048576)));
+
+        delegateOrcDataSource = new DelegateOrcDataSource(FakeOrcDataSource.INSTANCE);
+        cachingOrcDataSource = new CachingOrcDataSource(delegateOrcDataSource,
+                ImmutableList.of(new StripeInformation(123, 3, 10, 10, 10), new StripeInformation(123, 33, 10, 10, 10), new StripeInformation(123, 63, 1048576 * 8 - 20, 10, 10)));
+        cachingOrcDataSource.readCacheAt(62); // read at the end of a stripe
+        assertEquals(delegateOrcDataSource.getLastReadRanges(), ImmutableList.of(new DiskRange(33, 30)));
+        cachingOrcDataSource.readCacheAt(63);
+        assertEquals(delegateOrcDataSource.getLastReadRanges(), ImmutableList.of(new DiskRange(63, 8 * 1048576)));
+
+        delegateOrcDataSource = new DelegateOrcDataSource(FakeOrcDataSource.INSTANCE);
+        cachingOrcDataSource = new CachingOrcDataSource(delegateOrcDataSource,
+                ImmutableList.of(new StripeInformation(123, 3, 1, 0, 0), new StripeInformation(123, 4, 1048576, 1048576, 1048576 * 3), new StripeInformation(123, 4 + 1048576 * 5, 1048576, 1048576, 1048576)));
+        cachingOrcDataSource.readCacheAt(3);
+        assertEquals(delegateOrcDataSource.getLastReadRanges(), ImmutableList.of(new DiskRange(3, 1 + 1048576 * 5)));
+        cachingOrcDataSource.readCacheAt(4 + 1048576 * 5);
+        assertEquals(delegateOrcDataSource.getLastReadRanges(), ImmutableList.of(new DiskRange(4 + 1048576 * 5, 3 * 1048576)));
+
+        delegateOrcDataSource = new DelegateOrcDataSource(FakeOrcDataSource.INSTANCE);
+        cachingOrcDataSource = new CachingOrcDataSource(delegateOrcDataSource,
+                ImmutableList.of(new StripeInformation(123, 3, 1, 0, 0), new StripeInformation(123, 4, 1048576, 1048576, 1048576 * 3), new StripeInformation(123, 4 + 1048576 * 5, 1048576, 1048576, 1048576)));
+        cachingOrcDataSource.readCacheAt(4);
+        assertEquals(delegateOrcDataSource.getLastReadRanges(), ImmutableList.of(new DiskRange(4, 1048576 * 8)));
+    }
+
+    @Test
+    public void testIntegration()
+            throws IOException
+    {
+        DelegateOrcDataSource orcDataSource = new DelegateOrcDataSource(
+                new FileOrcDataSource(tempFile.getFile(), new DataSize(1, Unit.MEGABYTE), new DataSize(1, Unit.MEGABYTE), new DataSize(1, Unit.MEGABYTE)));
+        OrcReader orcReader = new OrcReader(orcDataSource, new OrcMetadataReader());
+        assertEquals(orcDataSource.getCount(), 1); // read file tail
+        List<StripeInformation> stripes = orcReader.getFooter().getStripes();
+        assertGreaterThanOrEqual(stripes.size(), 3); // Sanity check number of stripes. Normally is equal to 3. But can be 8 on Travis.
```

Comment:
"This can be three or higher, depending on ..."

And list the actual factor (CPU count, available memory, etc.) that causes it to vary rather than a specific test environment.

Why does the stripe count vary? Is this related to the Hive ORC memory manager?